bundles/org.openhab.binding.ecotouch/src/main/java/org/openhab/binding/ecotouch/internal/EcoTouchConnector.java

```diff
@@ -0,0 +1,387 @@
+import java.net.URLEncoder;
+import java.util.*;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Network communication with Waterkotte EcoTouch heat pumps.
+ *
+ * The communication protocol was reverse engineered from the Easy-Con Android
+ * app. The meaning of the EcoTouch tags was provided by Waterkotte's technical
+ * service (by an excerpt of a developer manual).
+ *
+ * @author Sebastian Held <sebastian.held@gmx.de> - Initial contribution
+ * @since 1.5.0
+ */
+
+public class EcoTouchConnector {
+    private String ip;
+    private String username;
+    private String password;
+    List<String> cookies;
+    static Pattern responsePattern = Pattern.compile("#(.+)\\s+S_OK[^0-9-]+([0-9-]+)\\s+([0-9-.]+)");
+
+    private final Logger logger = LoggerFactory.getLogger(EcoTouchConnector.class);
+
+    /**
+     * Create a network communication without having a current access token.
+     */
+    public EcoTouchConnector(String ip, String username, String password) {
+        this.ip = ip;
+        this.username = username;
+        this.password = password;
+        this.cookies = null;
+    }
+
+    /**
+     * Create a network communication with access token. This speeds up
+     * retrieving values, because the log in step is omitted.
+     */
+    public EcoTouchConnector(String ip, String username, String password, List<String> cookies) {
+        this.ip = ip;
+        this.username = username;
+        this.password = password;
+        this.cookies = cookies;
+    }
+
+    private synchronized void trylogin(boolean force) throws Exception {
+        logger.debug("trylogin: in");
+        if (force == false && cookies != null) {
+            logger.debug("trylogin: out fast");
+            return;
+        }
+        login();
+        logger.debug("trylogin: out normal");
+    }
+
+    private void login() throws Exception {
+        logger.debug("login");
+        cookies = null;
+        String url = null;
+        String line1 = null, line2 = null;
+        String cause = null;
+        try {
+            url = "http://" + ip + "/cgi/login?username=" + URLEncoder.encode(username, "UTF-8") + "&password="
+                    + URLEncoder.encode(password, "UTF-8");
+            URL loginurl = new URL(url);
+            URLConnection connection = loginurl.openConnection();
+            cookies = connection.getHeaderFields().get("Set-Cookie");
+            BufferedReader in = new BufferedReader(new InputStreamReader(connection.getInputStream()));
+            line1 = in.readLine();
+            line2 = in.readLine();
+        } catch (MalformedURLException e) {
+            cause = e.toString();
+            logger.debug("The URL '{}' is malformed: {}", url, cause);
+        } catch (Exception e) {
+            cause = e.toString();
+            logger.debug("Cannot log into Waterkotte EcoTouch: {}", cause);
+        }
```

Comment:
Should the thing go OFFLINE when these occur?

Reply:
Yeah, it does go OFFLINE, because the cookies could not be retrieved (no login token). It throws IOException() later